R/resample_data.R

```diff
@@ -28,83 +28,135 @@
 #'
 #'
 #' @export
+#'
+
 resample_data = function(data, N, ID_labels=NULL) {
-  # User didn't provide an N or an ID label, it's clear they just want a regular bootstrap
-  if (missing(N) & is.null(ID_labels)) {
-    N <- nrow(data)
-    return(bootstrap_single_level(data, nrow(data), ID_label=NULL))
-  }
+  # Mask internal outer_level and use_dt arguments from view.
+  .resample_data_internal(data, N, ID_labels)
+}
 
-  # Error handling
-  if (!is.null(ID_labels) & (length(N) != length(ID_labels))) {
-    stop(
-      "If you provide more than one ID_labels to resample data for multilevel data, please provide a vector for N of the same length representing the number to resample at each level."
-    )
-  }
+.resample_data_internal = function(data, N, ID_labels=NULL, outer_level=1, use_dt = NA) {
+  # Handle all the data sanity checks in outer_level so we don't have redundant error
+  # checks further down the recursion.
+  if(outer_level) {
+    # Optional usage of data.table to speed up functionality
+    # Short-circuit on the is.na to only attempt the package load if necessary.
+    if(is.na(use_dt) && requireNamespace("data.table", quietly=T)) {
+      use_dt = 1
+    } else {
+      use_dt = 0
+    }
```

Comment:
`use_dt <- use_dt || requireNamespace("data.table", quietly=T)`

Reply:
It wouldn't seem like this would work if we want a unit test to be able to explicitly override the value, no? If the unit test sets `use_dt = FALSE` in the argument, then the or operator is going to evaluate to `FALSE || requireNamespace("data.table", quietly=T)` which should evaluate to true.

Now, I could do this:
`use_dt = ifelse(is.na(use_dt), requireNamespace("data.table", quietly=T), use_dt)`

But it's not clear if speed or legibility benefit from this?